File: R/statistical_tests.r
#' Ordered Hypergeometric Test
#'
#' Perform a series of hypergeometric tests (a.k.a. Fisher's Exact tests), on a ranked list of genes ordered
#' by significance against a list of annotation genes. The hypergeometric tests are executed with 
#' increasingly larger numbers of genes representing the top genes in order of decreasing scores. 
#' The lowest p-value of the series is returned as the optimal enriched intersection of the ranked list of genes
#' and the biological term (pathway). 
#'
#' @param genelist Character vector of gene names, assumed to be ordered by decreasing importance. 
#' For example, the genes could be ranked by decreasing significance of differential expression. 
#' @param background Character vector of gene names. List of all genes used as a statistical background (i.e., the universe)
#' @param annotations Character vector of gene names. A gene set representing a functional term, process or biological pathway. 
#'
#' @return a list with the items:
#'   \describe{
#'     \item{p.val}{The lowest obtained p-value}
#'     \item{ind}{The index of \code{genelist} such that \code{genelist[1:ind]}
#'       gives the lowest p-value}
#'  }
#' @export
#'
#' @examples
#'    orderedHypergeometric(c('HERC2', 'SP100'), c('PHC2', 'BLM', 'XPC', 'SMC3', 'HERC2', 'SP100'),
#'                          c('HERC2', 'PHC2', 'BLM'))
orderedHypergeometric <- function(genelist, background, annotations) {
  # Only test subsets of genelist that end with a gene in annotations since
  # these are the only tests for which the p-value can decrease
  which.in <- which(genelist %in% annotations)
  if (length(which.in) == 0) return(list(p.val=1, ind=1))
  
  scores <- c()
  
  #go through all the lists and calculate the hyper geometric test
  for (i in 1:length(which.in)) {
    scores[i] <- 
      stats::phyper(
        q = i - 1, #overlap - 1 
        m = length(annotations),
        n = length(background) - length(annotations),
        k = which.in[i],
        lower.tail = FALSE)
  }
  
  # Return the lowest p-value and the associated index
  min.score <- min(scores)
  
  ind = which.in[max(which(scores==min.score))]
  p.val = min.score
  
  list(p.val=p.val, ind=ind)
}

File: Python/statistical_tests.py
import numpy as np
import scipy.stats as stats

def hypergeomtric(counts):
    if np.any(counts < 0):
        print("counts contains negative values. Something went very wrong.")
        return
    
    # NOTE check this
    m = counts[0,0] + counts[1,0]
    n = counts[0,1] + counts[1,1]
    k = counts[0,0] + counts[0,1]
    x = counts[0,0]

    return 1 - stats.hypergeom.cdf(x-1, n, m, k,loc=0)

def orderedHypergeometric(gene_arr, background, annotations):
    # Only test subsets of genelist that end with a gene in annotations since these are the only tests for which the p-value can decrease
    genes_indices = np.where(np.isin(gene_arr, annotations))[0]
    if len(genes_indices) == 0:
        return 1, 1

    gl = gene_arr[0:genes_indices[0]+1]
    cl = background[~(np.isin(background, gl))]
    genelist0 = len(gl) - 1
    complement1 = np.sum(np.isin(cl, annotations))
    complement0 = len(cl) - complement1

    counts = np.array([[1, complement1], [genelist0, complement0]])
    scores = [hypergeomtric(counts)]

    if len(genes_indices) == 1:
        return scores[0], genes_indices[0]


    for i in range(1,len(genes_indices)):
        diff = genes_indices[i] - genes_indices[i - 1]

        counts[0,0] = i + 1
        counts[1,0] = counts[1,0] + diff - 1
        counts[0,1] = counts[0,1] - 1
        counts[1,1] = counts[1,1] - diff + 1

        scores.append(hypergeomtric(counts))

    
    scores = np.array(scores).astype("float64")
    min_score = np.nanmin(scores)

    ind = genes_indices[np.max(np.where(scores == min_score))]
    p_val = min_score

    return p_val, ind
